### DISCUSSION TITLE
Navigating between blocks within a stack

### DISCUSSION THREAD

**Discussion by BeksOmega** [RaspberryPiFoundation/blockly-keyboard-experimentation#20] [General]
How might we enable users to navigate between blocks within a [stack](https://developers.google.com/blockly/guides/get-started/workspace-anatomy#block_stack)?

There are many different ways we could do this. We've 3 different models in the past, but the navigation actions within them are not mutually exclusive, and could be combined.

1) "Basic cursor". Uses pre order traversal to allow users to navigate through everything using only the previous and next command. This seems like it might work best for our limited motor control [Emily persona](https://github.com/google/blockly-keyboard-experimentation/wiki/Accessibility-Groups-and-Personas#emily-limited-motor-control-9-years-old), since it only use two keys. 
2) "Tree cursor" (aka default cursor). Uses previous, next, in, and out to navigate through the code as if it were an abstract syntax tree. This seems like it might work best for our blind [Alex persona](https://github.com/google/blockly-keyboard-experimentation/wiki/Accessibility-Groups-and-Personas#alex-blind-10-years-old), since it best represents the tree structure of the code.
3) "Line cursor". Tries to mimic a text editor. Navigating up and down will take the cursor to the next and previous "line" of code. Navigating in and out will move the cursor through all the fields and inputs in that "line" of code. This seems like it might work best for our low vision [Maria persona](https://github.com/google/blockly-keyboard-experimentation/wiki/Accessibility-Groups-and-Personas#maria-low-vision-15-years-old), and our RSI [Jason persona](https://github.com/google/blockly-keyboard-experimentation/wiki/Accessibility-Groups-and-Personas#jason-repetitive-stress-injury-17-years-old), since it reflects how you navigate similar applications like text editors.

One variant that has been proposed is providing basic line navigation, and also basic "in and out" navigation to get to different levels of the syntax tree. So this partially combines the "Tree cursor" and "Line cursor" implementations. **This is what is currently implemented.**

There may be other options not covered by these three implementations as well!

**Comment by BeksOmega** [comment]
I think this discussion is actually waiting on preliminary testing feedback (e.g. with friends/colleagues/family members) for the current basic line + in out proposal.

However alternate suggestions for schemes are also appreciated!

**Comment by samantha-code** [comment]
I completed some preliminary feedback with both younger students (1st grade classroom) and folks with accessibility needs. Although I only received three responses to my test plan, this is the output so far! The users were asked to rate each cursor on 5 tasks from very easy to very difficult. The five tasks were: 

- Task 1: Select the block from the toolbox and add it under the event block in the workspace. 
- Task 2: Select the block from the toolbox and add it between two blocks inside of a Repeat/loop block
-Task 3: Task 3: Delete a block inside of the repeat/loop block
- Moving between block stacks 
-Moving within a block stack

Below is the outcome of the testing - hope this helps further the discussion a bit
[Keyboard Navigation Testing Results.pdf](https://github.com/user-attachments/files/16236943/Keyboard.Navigation.Testing.Results.pdf)

**Reply by BeksOmega** [reply]
Interesting! Thanks for sharing that @samantha-code :D It looks like the folks that responded were leaning toward the basic/tree cursor. Do you know if these users were sighted? I'm curious if what we're seeing aligns with the hypotheses in the OP about the basic/tree cursor being better for blind users.
